### ISSUE TITLE
SVGElement *A = (SVGElement *)B,B is kind of class Element, demo crashed!

### ISSUE THREAD

**Issue by sun2015** [SVGKit/SVGKit#348]
Recently,  i use svgkit to draw complex svg image. Unfortunately,demo crash. 
I found this:
SVGElement\* linkedElement = (SVGElement*) [parseResult.parsedDocument getElementById:linkHref];

in other word, subclass  SVGElement = (SVGElement *)superclass Element. it's wrong.

Do you have any idea to solve the problem?

**Comment by adamgit** [comment]
As per the guidelines, you have to show us the lines of your SVG. We cannot 
read your mind.

**Comment by sun2015** [comment]
@adamgit  i am sorry.
 First, i run Demo-IOS.xcodeproj,replacing a svg image in local file with a my svg image. Then, demo crash in SVGKImage.m. 
- (CALayer *)newLayerWithElement:(SVGElement <ConverterSVGToCALayer> *)element
  {
  ...
  
  if( [element isKindOfClass:[SVGUseElement class]] )
  {
      SVGUseElement\* useElement = (SVGUseElement_) element;
      element = (SVGElement <ConverterSVGToCALayer>   *)useElement.instanceRoot.correspondingElement;  
  ...
  }
  ...
  NSString_ clipPath = [element cascadedValueForStylableProperty:@"clip-path" inherit:NO];
  ...
  }

the error is:
2015-07-08 09:31:10.880 Demo-iOS[1912:46490] -[Element cascadedValueForStylableProperty:inherit:]: unrecognized selector sent to instance 0x7989aa90

i find the error is because of the statement (element = (SVGElement <ConverterSVGToCALayer>   *)useElement.instanceRoot.correspondingElement;).

then, continue to debug the project.

Finally,  in  SVGKParserDefsAndUse.m 
- (Node_) handleStartElement:(NSString *)name document:(SVGKSource_) SVGKSource namePrefix:(NSString_)prefix namespaceURI:(NSString_) XMLNSURI attributes:(NSMutableDictionary _)attributes parseResult:(SVGKParseResult *)parseResult parentNode:(Node_) parentNode
  {
  ... 
  
  /*\* have to find the node in the DOM tree with id = xlink:href's value _/
  SVGElement_ linkedElement = (SVGElement*) [parseResult.parsedDocument getElementById:linkHref];  
  useElement.instanceRoot = [self convertSVGElementToElementInstanceTree:linkedElement outermostUseElement:useElement];
  ...
  }

In my view, the statement: SVGElement\* linkedElement = (SVGElement*) [parseResult.parsedDocument getElementById:linkHref]; is wrong.The statement converse superclass Element to subclass  SVGElement, but the conversion is false.

Now, i need a good idea to solve the crash. Thanks!

**Comment by adamgit** [comment]
Should it be Element, though? I think it should be SVGElement. Originally,
it always returned an SVGElement, but recent changes have - apparently -
changed that (to fix other issues). This might be a mistake.

What's the line of the SVG file causing this?

It appears to be a "use" tag, but what is it pointing to within the SVG
file, and why is this different from the SVG files we have in the demo
project that use the "use" tag and work OK?

The problem appears to be happening somewhere in this line:

   useElement.instanceRoot = [self
convertSVGElementToElementInstanceTree:linkedElement
outermostUseElement:useElement];

... that "convertSVGElementToElementInstanceTree" is - MAYBE - returning
something that is NOT an SVGElement, but which should be.

**Comment by sun2015** [comment]
@adamgit 
when debug the project, the line:
SVGElement linkedElement = (SVGElement*) [parseResult.parsedDocument getElementById:linkHref];
show that linkedElement is an Element,not SVGElement. 

In SVG file, the "use" tag like this:
![2015-07-09 3 16 04](https://cloud.githubusercontent.com/assets/12063567/8589885/ae60d5ca-264d-11e5-89fe-bcf70b750c44.png)

![2015-07-09 3 14 40](https://cloud.githubusercontent.com/assets/12063567/8589873/8a7aa8ac-264d-11e5-97da-8d5600042613.png)

**Comment by jtmilne** [comment]
I am getting the same issue when trying to load my svg. 

```
2015-08-13 22:05:33.448 SBSeatMapView[13104:883023] -[Element cascadedValueForStylableProperty:inherit:]: unrecognized selector sent to instance 0x7f87db6c8a10

2015-08-13 22:05:33.458 SBSeatMapView[13104:883023] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[Element cascadedValueForStylableProperty:inherit:]: unrecognized selector sent to instance 0x7f87db6c8a10'

*** First throw call stack:
(
0   CoreFoundation                      0x0000000103182c65 __exceptionPreprocess + 165
1   libobjc.A.dylib                     0x0000000102e1bbb7 objc_exception_throw + 45
2   CoreFoundation                      0x000000010318a0ad -[NSObject(NSObject) doesNotRecognizeSelector:] + 205
3   CoreFoundation                      0x00000001030e013c ___forwarding___ + 988
4   CoreFoundation                      0x00000001030dfcd8 _CF_forwarding_prep_0 + 120
5   SBSeatMapView                       0x0000000100ab08a3 -[SVGKImage newLayerWithElement:] + 755
6   SBSeatMapView                       0x0000000100ab1130 -[SVGKImage newLayerWithElement:] + 2944
7   SBSeatMapView                       0x0000000100ab1130 -[SVGKImage newLayerWithElement:] + 2944
8   SBSeatMapView                       0x0000000100ab1130 -[SVGKImage newLayerWithElement:] + 2944
9   SBSeatMapView                       0x0000000100ab1364 -[SVGKImage newCALayerTree] + 132
10  SBSeatMapView                       0x0000000100ab1776 -[SVGKImage CALayerTree] + 294
11  SBSeatMapView                       0x0000000100ab5122 -[SVGKLayeredImageView initWithSVGKImage:] + 226
.....
```

**Comment by adamgit** [comment]
@jtmilne stacktrace doesn't really tell people anything. You need to show what line of SVG triggered it, and / or look in your debugger and tell us what you see.

**Comment by adamgit** [comment]
@sun2015 I think it was @dgileadi who made the changes with "use" tags - he might have some more thoughts to add here?

**Comment by sun2015** [comment]
@adamgit I will update the latest verson,thanks

**Comment by dgileadi** [comment]
@adamgit I don't recall changing "use" tags. Looking through the pull requests I see some "use" changes by @greghe; maybe he is more familiar with them?

**Comment by jtmilne** [comment]
I will dig in and try and pinpoint what element of the svg is causing this. However, I did checkout a series of random commits going back about a year and none of them worked, so in my case this is not something that was introduced recently.

**Comment by adamgit** [comment]
Ah, well ... USE never worked at all, originally (I did the initial
version, and although I wrote all the code, I couldn't get it to work).

Whoever fixed that - @greghe perhaps (sorry, @dgileadi!) - found a few
small stupid mistakes in my code that seemed to suddenly make it work, and
added some improvements too.

BUT ... some of the improvements I remember changed the hierarchy of
SVGElement / Element - and that's what I'm suspicious has gone wrong here.
